### PR TITLE
[FSDP] Add exec order validation

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#75902 [FSDP] Add exec order validation**

This adds execution order validation for `FSDP`.

**Background**
For collectives, we may classify three types of mismatches:
1. Collective type mismatch: Rank `i` schedules one collective type while rank `j` schedules another collective type.
	- This errors immediately.
2. Collective shape mismatch: Ranks schedule the same collective but one rank's contribution has a mismatched shape.
	- This errors immediately.
3. Logical collective mismatch: Collective type and shapes match but rank `i`'s collective should not logically match with rank `j`'s collective.
	- This error silently.

Our goal with this execution order validation is to improve the user experience around class-3 mismatches. `TORCH_DISTRIBUTED_DEBUG=DETAIL` may help with class-1 and class-2 mismatches (not implemented yet).

**Approach**
Assumption 1: The `FSDP` model `model` is wrapped identically across ranks. This allows each rank to establish a common `param` to `param_index` bijection (by enumerating `model.parameters()`).
Assumption 2: If the backward pass all-gathers are in order, then the backward pass reduce-scatters are in order. Hence, this approach only checks all-gathers.

- First iteration:
    - Before every all-gather (in `_rebuild_full_params()`), we all-gather a one-dimension `torch.tensor(param_index)` from each rank indicating the parameter the rank is about to all-gather. All `param_index` s should be the same.
- Subsequent iterations:
	- We record the parameter order. If the order deviates from that of the first iteration, then we issue a *warning* to the user.

Differential Revision: [D35684277](https://our.internmc.facebook.com/intern/diff/D35684277)

### CODE REVIEW

**Comment by awgu** [comment]
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35684277).

**Comment by awgu** [comment]
Failures are real. Working on it...

**Comment by awgu** [comment]
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35684277).

**Comment by awgu** [comment]
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35684277).

File: torch/distributed/fsdp/fully_sharded_data_parallel.py
Comment (awgu) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:1856]: I moved this state transition to earlier in the pre-backward hook since otherwise the call to `_rebuild_full_params()` is still under `IDLE` rather than `PRE_BACKWARD`. The training state should not affect the execution of any of the code between this new position and the old position a few lines down. However, it could be good to get a second pair of eyes to sanity check this.

File: torch/distributed/fsdp/fully_sharded_data_parallel.py
Comment (awgu) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:2401]: I am doing this pairwise check so that the error message can concretely include where execution orders differ (i.e. which rank and which index).

File: torch/distributed/fsdp/fully_sharded_data_parallel.py
Comment (awgu) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:2364]: The previous commit in the stack was important to ensuring this.

File: test/distributed/fsdp/test_fsdp_exec_order.py
Comment (awgu) [inline test/distributed/fsdp/test_fsdp_exec_order.py:140]: I did not include "valid" tests since the existing test suite are all "valid". I did not see any false positive warnings when running through `test_fsdp_core.py`, but please feel free to let me know if after this is landed, we start seeing warnings for other tests.

**Review by awgu — COMMENTED** [review]
Some comments to help review

**Comment by awgu** [comment]
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35684277).

File: torch/distributed/fsdp/fully_sharded_data_parallel.py
Comment (rohan-varma) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:2397]: I think this incurs extra CPU -> GPU copy, can we do torch.zeros(world_size, dtype=torch.int32, device=device)

File: torch/distributed/fsdp/fully_sharded_data_parallel.py
Comment (rohan-varma) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:2399]: what happens if user is gathering the same parameter index but the module somehow has different shape? Does broadcast at the beginning or torch_distributed_debug anticipated to help with this?

File: torch/distributed/fsdp/fully_sharded_data_parallel.py
Comment (rohan-varma) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:2404]: can't we just do `if not torch.equal(...)` and raise the error directly?

File: test/distributed/fsdp/test_fsdp_exec_order.py
Comment (rohan-varma) [reply test/distributed/fsdp/test_fsdp_exec_order.py:133]: Wonder if the warning will be hit every single iteration, will this become a bit too verbose for users who change the module in the same way across all ranks?

File: torch/distributed/fsdp/fully_sharded_data_parallel.py
Comment (rohan-varma) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:263]: It looks like `param_order` also gets modified, do we not need to reset this?

File: torch/distributed/fsdp/fully_sharded_data_parallel.py
Comment (rohan-varma) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:667]: Does `.parameters()` only contain `FlatParameter` in all cases? Is there a reason we can't use `self.params`?

File: torch/distributed/fsdp/fully_sharded_data_parallel.py
Comment (rohan-varma) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:1856]: Hmmm, I guess it depends on whether we want to consider this all_gather as part of backward_pre or not?

**Review by rohan-varma — APPROVED** [review]
LGTM, just some minor questions / concerns for your consideration. @zhaojuanmao can also take a look, thank you!

File: torch/distributed/fsdp/fully_sharded_data_parallel.py
Comment (awgu) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:241]: Yeah. This is just simply enumerating the flattened parameters. Another approach would have been to `enumerate(self.parameters())` every time, but I decided to save it here.

File: torch/distributed/fsdp/fully_sharded_data_parallel.py
Comment (awgu) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:2336]: Could you clarify the confusion with the name? I meant that it is checking the all-gather in the next line, not that the check itself runs an all-gather.

File: torch/distributed/fsdp/fully_sharded_data_parallel.py
Comment (awgu) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:2404]: Yes. I included to the `continue` so that the error message has one fewer indent, but I can error directly if that is preferred.

File: torch/distributed/fsdp/fully_sharded_data_parallel.py
Comment (awgu) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:667]: `self.params` only contains the root module's parameters. `.parameters()` includes the `FlatParameter`s of all nested FSDP modules as well.

File: torch/distributed/fsdp/fully_sharded_data_parallel.py
Comment (awgu) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:2399]: I agree then it is problematic. I guess with your diff for checking parameter shapes, this is not an issue since the shapes will already be validated.

File: test/distributed/fsdp/test_fsdp_exec_order.py
Comment (awgu) [reply test/distributed/fsdp/test_fsdp_exec_order.py:138]: Yes, this will be a false positive. However, since the warning message is written to say that a change has been detected, not necessarily that something is wrong, hopefully it is not too confusing for the user.

File: torch/distributed/fsdp/fully_sharded_data_parallel.py
Comment (zhaojuanmao) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:2385]: could we give Torch_Warning_Once()?

Comment (zhaojuanmao) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:2385]: also does this need to return here? otherwise it will crash if it calls eod.param_order[eod.index] in the next line?

File: torch/distributed/fsdp/fully_sharded_data_parallel.py
Comment (zhaojuanmao) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:2388]: Torch_Warning_Once?

**Review by zhaojuanmao — APPROVED** [review]
overall looks great! some minor comments. Like how it handles mismatch before hang!

**Comment by zhaojuanmao** [comment]
forgot to mention that we should run benchmarks before landing this PR, as it touched core codes

File: torch/distributed/fsdp/fully_sharded_data_parallel.py
Comment (awgu) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:2377]: This `_check_all_gather()` prefaces the all-gather in `_rebuild_full_params()`. However, we may call `_rebuild_full_params()` in training states other than `FORWARD` and `BACKWARD_PRE`, in which case we do not want to include this check. (For example, we call `_rebuild_full_params()` in `summon_full_params()`.)

File: torch/distributed/fsdp/fully_sharded_data_parallel.py
Comment (awgu) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:263]: Yes, `param_to_param_index` is initialized once and never changed, and param_order` is constructed over the course of the first iteration and not changed thereafter.

File: test/distributed/fsdp/test_fsdp_exec_order.py
Comment (awgu) [reply test/distributed/fsdp/test_fsdp_exec_order.py:133]: Changing this to warn only once per rank.

File: torch/distributed/fsdp/fully_sharded_data_parallel.py
Comment (rohan-varma) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py]: awesome!

**Comment by awgu** [comment]
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35684277).

**Comment by awgu** [comment]
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35684277).

**Comment by awgu** [comment]
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35684277).

**Comment by awgu** [comment]
I updated the internal diff with the relevant benchmark flows. To be fully transparent, I do see around 1-2% P50 QPS regression in the 1 node 8 GPU setting (consistently across 3 runs for both RoBERTa and ResNet).

I also made some non-trivial code changes (though the overall logic is still generally the same), so I am re-requesting review.

File: torch/distributed/fsdp/fully_sharded_data_parallel.py
Comment (zhaojuanmao) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:2500]: nit: do we want to warn only once at the first time when the mismatch is found? because the main purpose is for users to notice this change, possibly one warning is sufficient?

**Review by zhaojuanmao — APPROVED** [review]

File: torch/distributed/fsdp/fully_sharded_data_parallel.py
Comment (awgu) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:2500]: From earlier offline discussion in the devs chat, I think it is still good to include warnings through the first deviating iteration in case the order change is somehow unintentional. Then, providing some additional information may be helpful for debugging.
I think it should be fine since the messages are still contained to be for one iteration and will not flood the console constantly.

**Pull request closed without merging**
